flink-rpc/flink-rpc-akka/src/main/java/org/apache/flink/runtime/rpc/akka/RobustActorSystem.java

```diff
@@ -85,17 +86,52 @@ private static RobustActorSystem create(
                                 .map(BootstrapSetup::defaultExecutionContext)
                                 .flatMap(RobustActorSystem::toJavaOptional));
 
+        final PostShutdownClassLoadingErrorFilter postShutdownClassLoadingErrorFilter =
+                new PostShutdownClassLoadingErrorFilter(uncaughtExceptionHandler);
+
         final RobustActorSystem robustActorSystem =
                 new RobustActorSystem(name, appConfig, classLoader, defaultEC, setup) {
                     @Override
                     public Thread.UncaughtExceptionHandler uncaughtExceptionHandler() {
-                        return uncaughtExceptionHandler.getOrElse(super::uncaughtExceptionHandler);
+                        return postShutdownClassLoadingErrorFilter;
                     }
                 };
+        robustActorSystem.registerOnTermination(
+                postShutdownClassLoadingErrorFilter::notifyShutdownComplete);
+
         robustActorSystem.start();
         return robustActorSystem;
     }
 
+    private static class PostShutdownClassLoadingErrorFilter
+            implements Thread.UncaughtExceptionHandler {
+
+        private final AtomicBoolean shutdownComplete = new AtomicBoolean();
+        private final Thread.UncaughtExceptionHandler uncaughtExceptionHandler;
+
+        public PostShutdownClassLoadingErrorFilter(
+                Thread.UncaughtExceptionHandler uncaughtExceptionHandler) {
+            this.uncaughtExceptionHandler = uncaughtExceptionHandler;
+        }
+
+        public void notifyShutdownComplete() {
+            shutdownComplete.set(true);
+        }
+
+        @Override
+        public void uncaughtException(Thread t, Throwable e) {
+            if (shutdownComplete.get()
+                    && (e instanceof NoClassDefFoundError || e instanceof ClassNotFoundException)) {
```

Comment:
I'm wondering whether we should make this more restrictive (checking that the thread's name includes the `akka` keyword). So far, we have seen only `flink-akka` and `flink-metrics-akka` as the current thread in issues reported under FLINK-32189.

Reply:
I can do that, but this uncaught exception handler only applies to the akka actor system, so it should implicitly only cover akka threads (independent of how they are named).

Reply:
ah ok, I actually wanted to check where the class is used but forgot. You're right - it doesn't make a difference. then.